### PR TITLE
Add SDK caching & Refactor

### LINKED ISSUE #74: Autoscaling prevents ASG from rolling update.
**Is this a BUG REPORT or FEATURE REQUEST?**:

BUG

**What happened**:
If ASGs was scaled in during rolling upgrade, upgrade will fail.

**What you expected to happen**:
https://github.com/keikoproj/upgrade-manager/blob/4f02093e60cd739a5966c5de687c368efa08e660/controllers/rollingupgrade_controller.go#L297-L319 - should refresh ASG details.

**How to reproduce it (as minimally and precisely as possible)**:
Start rollout, then manually scale down the ASG.

**Anything else we need to know?**:

**Environment**:
- rolling-upgrade-controller version
- Kubernetes version : 1.16.7

**Other debugging information (if applicable)**:
- RollingUpgrade status:
```
$ kubectl describe rollingupgrade <rollingupgrade-name>
```
- controller logs:
```
$ kubectl logs <rolling-upgrade-controller pod>

upgrade-manager-7c56c9d469-xgf6c rolling-upgrade-controller time="2020-03-10T21:29:24Z" level=info msg="rollingupgrade-node-eu-central-1b: desired capacity is met, 22 instances in service"
upgrade-manager-7c56c9d469-xgf6c rolling-upgrade-controller time="2020-03-10T21:29:24Z" level=info msg="rollingupgrade-node-eu-central-1b: desired capacity is met, 22 instances in service"
upgrade-manager-7c56c9d469-xgf6c rolling-upgrade-controller time="2020-03-10T21:29:24Z" level=info msg="rollingupgrade-node-eu-central-1b: new node has not yet joined the cluster"
...
upgrade-manager-7c56c9d469-xgf6c rolling-upgrade-controller time="2020-03-10T21:39:21Z" level=info msg="rollingupgrade-node-eu-central-1b: new node has not yet joined the cluster"
upgrade-manager-7c56c9d469-xgf6c rolling-upgrade-controller time="2020-03-10T21:39:22Z" level=info msg="Deleted the entries of ASG XXX in the cluster store for rollingupgrade-node-eu-central-1b"
upgrade-manager-7c56c9d469-xgf6c rolling-upgrade-controller time="2020-03-10T21:39:22Z" level=info msg="Marked object rollingupgrade-node-eu-central-1b as error"
```

### PR DESCRIPTION
**Pull request by eytan-avisror**
Fixes #74 

Cache will expire once a minute, and be flushed whenever writes happen (Terminate/SetStandby/etc).

Added the required call to populateASG

## Refactor

The refactoring removes the region from spec and makes it a controller option controlled by a flag `--region`, this also deprecates the API option for `.spec.region` and removes all references to it.

SDK instantiation is moved to main.go in order to not mess up unit tests with caching

### Testing

- [x] Scale down does not fail rollup

#### while rollup was running on 1 node, killed other 2 in scaling group
```
time="2020-03-10T23:31:13Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:31:13Z" level=info msg="test-rollup: new instance has not yet joined the scaling group"
time="2020-03-10T23:32:44Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:32:44Z" level=info msg="test-rollup: desired capacity is met, 3 instances in service"
time="2020-03-10T23:32:44Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:32:44Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:34:14Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:34:14Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:34:59Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:34:59Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:35:21Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:35:21Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:35:36Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:35:36Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:35:51Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:35:51Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:36:06Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:36:06Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:36:21Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:36:21Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:36:36Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:36:36Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:36:51Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:36:51Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:37:06Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:37:06Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:37:22Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:37:22Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:37:37Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:37:37Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:37:52Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:37:52Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:38:07Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:38:07Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:38:22Z" level=debug msg="cache hit => false, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:38:22Z" level=info msg="test-rollup: new node has not yet joined the cluster"
time="2020-03-10T23:38:37Z" level=debug msg="cache hit => true, service => autoscaling.DescribeAutoScalingGroups"
time="2020-03-10T23:38:37Z" level=info msg="test-rollup: desired capacity is met, 3 nodes in service"
```

- [x] Many ASGs do not get throttled excessively

### CODE REVIEW

**Comment by uthark** [comment]
Need `Action required` section in release notes/upgrade instructions.

**Comment by eytan-avisror** [comment]
> Need `Action required` section in release notes/upgrade instructions.

Yep, will be added in release notes of the actual release

File: main.go [hunk 1]
```diff
@@ -73,10 +102,40 @@ func main() {
 		os.Exit(1)
 	}
 
+	if region == "" {
+		setupLog.Error(err, "--region flag not provided")
```
Comment (uthark) [inline main.go]: Shouldn't it also support discovery via ec2 metadata API?

Comment (eytan-avisror) [reply main.go]: Yes, but that would make running locally `make run` problematic

Comment (eytan-avisror) [reply main.go]: We will leverage environment variable - I will fix it

**Comment by eytan-avisror** [comment]
@shrinandj can you take a look?

**Comment by uthark** [comment]
@eytan-avisror waiting for this to be merged.

**Pull request merged**